File: C++/src/Tank.cpp
// src/Tank.cpp
#include "Tank.h"
#include "Utils.h"
#include "ConfigManager.h"
#include "resource/TextureManager.h"
#include "renderer/Renderer.h"
#include <cmath>

// ===== 构造函数 =====
Tank::Tank(float x, float y, sf::Color color, float speed, bool isPlayer, int pid)
    : x(x), y(y), w(getTankSize()), h(getTankSize()),
      color(color), speed(speed),
      player(isPlayer), playerId(pid), dirX(0), dirY(-1),
      lives(isPlayer ? getPlayerLives() : 1), cooldown(0), alive(true), frameCounter(0) {}

// ===== update =====
void Tank::update(float dt) {
    if (cooldown > 0) cooldown -= dt;

    for (auto it = effects.begin(); it != effects.end(); ) {
        it->second -= dt;
        if (it->second <= 0) {
            it = effects.erase(it);
        } else {
            ++it;
        }
    }

    frameCounter++;
    if (frameCounter % 3 == 0 && alive) {
        TrailPoint tp;
        tp.x = x + w/2;
        tp.y = y + h/2;
        tp.age = 0;
        trailPoints.push_back(tp);
    }

    for (auto it = trailPoints.begin(); it != trailPoints.end(); ) {
        it->age += dt;
        if (it->age >= 1.0f) {
            it = trailPoints.erase(it);
        } else {
            ++it;
        }
    }

    for (auto it = bullets.begin(); it != bullets.end(); ) {
        it->update(dt);
        if (!it->isAlive()) {
            it = bullets.erase(it);
        } else {
            ++it;
        }
    }
}

// ===== draw =====
void Tank::draw(sf::RenderWindow& window) {
    // ===== 痕迹 =====
    for (const auto& tp : trailPoints) {
        float alpha = 60.0f * (1.0f - tp.age / 1.0f);
        if (alpha > 5.0f) {
            sf::RectangleShape rect(sf::Vector2f(w, h));
            rect.setPosition(sf::Vector2f(tp.x - w/2.0f, tp.y - h/2.0f));
            rect.setFillColor(sf::Color(0, 0, 0, static_cast<uint8_t>(alpha)));
            window.draw(rect);
        }
    }

    if (!alive) return;

    float scale = getSizeScale();
    int wDraw = static_cast<int>(w * scale);
    int hDraw = static_cast<int>(h * scale);
    int xDraw = static_cast<int>(x - (wDraw - w) / 2.0f);
    int yDraw = static_cast<int>(y - (hDraw - h) / 2.0f);

    std::string entityId;
    if (player) {
        entityId = "tank_p" + std::to_string(playerId);
    } else if (isBoss) {
        entityId = "tank_boss";
    } else {
        entityId = "tank_enemy";
    }

    Renderer& renderer = Renderer::getInstance();

    // ===== 尝试使用贴图 =====
    if (renderer.drawTankWithTexture(window, this, entityId, wDraw, hDraw, xDraw, yDraw)) {
        // 玩家编号（叠加）
        if (player) {
            sf::Font font;
            if (font.openFromFile("assets/fonts/arial.ttf") ||
                font.openFromFile("assets/fonts/consola.ttf")) {
                sf::Text text(font, std::to_string(playerId), 16);
                text.setFillColor(sf::Color::Black);
                text.setStyle(sf::Text::Bold);
                sf::Vector2f center = getCenter();
                sf::FloatRect bounds = text.getLocalBounds();
                text.setPosition(sf::Vector2f(
                    center.x - bounds.size.x / 2.0f,
                    center.y - bounds.size.y / 2.0f - 2.0f
                ));
                window.draw(text);
            }
        }
        return;
    }

    // ===== 无贴图：使用内置绘制 =====
    renderer.drawTankBuiltin(window, this, color, wDraw, hDraw, xDraw, yDraw);

    // 玩家编号
    if (player) {
        sf::Font font;
        if (font.openFromFile("assets/fonts/arial.ttf") ||
            font.openFromFile("assets/fonts/consola.ttf")) {
            sf::Text text(font, std::to_string(playerId), 14);
            text.setFillColor(sf::Color::Black);
            text.setStyle(sf::Text::Bold);
            sf::Vector2f center = getCenter();
            sf::FloatRect bounds = text.getLocalBounds();
            text.setPosition(sf::Vector2f(
                center.x - bounds.size.x / 2.0f,
                center.y - bounds.size.y / 2.0f - 2.0f
            ));
            window.draw(text);
        }
    }

    // Boss 血条
    if (isBoss) {
        Renderer::getInstance().drawBossHealthBar(window, this, maxHp);
    }
}

// ===== drawBossHealthBar =====
void Tank::drawBossHealthBar(sf::RenderWindow& window) {
    if (!isBoss || maxHp <= 0) return;

    sf::Vector2f center = getCenter();
    int barWidth = w + 20;
    int barHeight = 8;
    int barX = static_cast<int>(center.x - barWidth / 2.0f);
    int barY = static_cast<int>(y - 16);

    float hpRatio = static_cast<float>(lives) / maxHp;

    sf::RectangleShape barBg(sf::Vector2f(barWidth, barHeight));
    barBg.setPosition(sf::Vector2f(barX, barY));
    barBg.setFillColor(sf::Color(40, 40, 50));
    window.draw(barBg);

    sf::Color hpColor;
    if (hpRatio > 0.5f) hpColor = sf::Color(0, 200, 0);
    else if (hpRatio > 0.25f) hpColor = sf::Color(200, 200, 0);
    else hpColor = sf::Color(200, 50, 50);

    sf::RectangleShape barHp(sf::Vector2f(
        static_cast<int>(barWidth * hpRatio),
        barHeight
    ));
    barHp.setPosition(sf::Vector2f(barX, barY));
    barHp.setFillColor(hpColor);
    window.draw(barHp);

    sf::RectangleShape border(sf::Vector2f(barWidth, barHeight));
    border.setPosition(sf::Vector2f(barX, barY));
    border.setFillColor(sf::Color::Transparent);
    border.setOutlineColor(sf::Color(255, 215, 0));
    border.setOutlineThickness(1.0f);
    window.draw(border);
}

// ===== move =====
void Tank::move(int dx, int dy, std::vector<Wall>& walls) {
    if (dx == 0 && dy == 0) return;
    
    float currentSpeed = speed * getSpeedMultiplier();
    float step = MOVE_STEP;
    float stepDx = dx * step;
    float stepDy = dy * step;
    int totalSteps = static_cast<int>(currentSpeed / step);
    if (totalSteps < 1) totalSteps = 1;

    for (int i = 0; i < totalSteps; ++i) {
        float newX = x + stepDx;
        float newY = y + stepDy;
        int margin = 2;
        float maxX = WINDOW_WIDTH - w - margin;
        float maxY = WINDOW_HEIGHT - h - margin;
        newX = clampf(newX, margin, maxX);
        newY = clampf(newY, margin, maxY);

        sf::FloatRect testRect(sf::Vector2f(newX, newY), sf::Vector2f(w, h));
        bool blocked = false;
        for (auto& wall : walls) {
            if (wall.isAlive() && testRect.findIntersection(wall.getRect()).has_value()) {
                blocked = true;
                break;
            }
        }
        
        if (!blocked) {
            x = newX;
            y = newY;
        } else {
            if (dx != 0) {
                sf::FloatRect testX(sf::Vector2f(newX, y), sf::Vector2f(w, h));
                bool blockedX = false;
                for (auto& wall : walls) {
                    if (wall.isAlive() && testX.findIntersection(wall.getRect()).has_value()) {
                        blockedX = true;
                        break;
                    }
                }
                if (!blockedX) x = newX;
            }
            if (dy != 0) {
                sf::FloatRect testY(sf::Vector2f(x, newY), sf::Vector2f(w, h));
                bool blockedY = false;
                for (auto& wall : walls) {
                    if (wall.isAlive() && testY.findIntersection(wall.getRect()).has_value()) {
                        blockedY = true;
                        break;
                    }
                }
                if (!blockedY) y = newY;
            }
            break;
        }
    }
}

// ===== shoot =====
Bullet* Tank::shoot() {
    if (cooldown > 0) return nullptr;
    if (bullets.size() >= MAX_BULLETS) return nullptr;

    cooldown = SHOT_COOLDOWN;
    sf::Vector2f firePoint = getFirePoint();

    float bulletSpeed = BULLET_SPEED * getBulletSpeedMultiplier();
    int bulletSize = static_cast<int>(BULLET_SIZE * getSizeScale());
    int bulletDamage = getBulletDamage();

    sf::Color bulletColor;
    if (player) {
        bulletColor = (playerId == 1) ? COLOR_BULLET_P1 : COLOR_BULLET_P2;
    } else {
        bulletColor = COLOR_BULLET_ENEMY;
    }

    bullets.emplace_back(firePoint.x, firePoint.y, dirX, dirY, player, playerId, bulletColor);
    Bullet& bullet = bullets.back();
    bullet.speed = bulletSpeed;
    bullet.w = bulletSize;
    bullet.h = bulletSize;
    bullet.damage = bulletDamage;
    return &bullet;
}

// ===== getRect =====
sf::FloatRect Tank::getRect() const {
    return sf::FloatRect(sf::Vector2f(x, y), sf::Vector2f(w, h));
}

// ===== getCenter =====
sf::Vector2f Tank::getCenter() const {
    return sf::Vector2f(x + w/2.0f, y + h/2.0f);
}

// ===== getFirePoint =====
sf::Vector2f Tank::getFirePoint() const {
    sf::Vector2f center = getCenter();
    float offset = w / 2.0f + 2.0f;
    return sf::Vector2f(center.x + dirX * offset, center.y + dirY * offset);
}

File: Java/src/com/tankbattle/Tank.java
package com.tankbattle;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Iterator;
import java.util.Map;
import java.util.HashMap;

import static com.tankbattle.Utils.*;

public class Tank {
    public double x, y;
    public int w, h;
    public Color color;
    public double speed;
    public boolean isPlayer;
    public int playerId;
    public Utils.Vec2 dir;
    public int lives;
    public double cooldown;
    public boolean alive;
    public List<Bullet> bullets;
    public java.util.Map<String, Float> effects = new java.util.HashMap<>();

    // ===== 行驶痕迹 =====
    private ArrayList<TrailPoint> trailPoints;
    private int frameCounter = 0;

    private class TrailPoint {
        double x, y, age;
    }

    // ===== 道具辅助方法 =====
    public boolean isProtected() {
        return effects.containsKey("protection");
    }

    public float getSpeedMultiplier() {
        return effects.containsKey("speed") ? 1.5f : 1.0f;
    }

    public float getBulletSpeedMultiplier() {
        return effects.containsKey("speed") ? 1.5f : 1.0f;
    }

    public float getSizeScale() {
        return effects.containsKey("strength") ? 1.5f : 1.0f;
    }

    public int getBulletDamage() {
        return effects.containsKey("strength") ? 2 : 1;
    }

    public Tank(double x, double y, Color color, double speed, boolean isPlayer, int playerId) {
        this.x = x;
        this.y = y;
        this.w = TANK_SIZE;
        this.h = TANK_SIZE;
        this.color = color;
        this.speed = speed;
        this.isPlayer = isPlayer;
        this.playerId = playerId;
        this.dir = new Utils.Vec2(0, -1);
        this.lives = isPlayer ? PLAYER_LIVES : 1;
        this.cooldown = 0;
        this.alive = true;
        this.bullets = new ArrayList<>();
        this.trailPoints = new ArrayList<>();
    }

    public Rectangle getRect() {
        return new Rectangle((int) x, (int) y, w, h);
    }

    public Utils.Vec2 getCenter() {
        return new Utils.Vec2(x + w / 2.0, y + h / 2.0);
    }

    public Utils.Vec2 getFirePoint() {
        Utils.Vec2 center = getCenter();
        double offset = w / 2.0 + 2;
        return new Utils.Vec2(center.x + dir.x * offset, center.y + dir.y * offset);
    }

    public void move(int dx, int dy, java.util.List<Wall> walls) {
        if (dx == 0 && dy == 0) return;

        // 应用 Speed 效果
        float speedMultiplier = effects.containsKey("speed") ? 1.5f : 1.0f;
        float currentSpeed = (float)(this.speed * speedMultiplier);

        double step = MOVE_STEP;
        double stepDx = dx * step;
        double stepDy = dy * step;

        int totalSteps = (int)(currentSpeed / step);
        if (totalSteps < 1) totalSteps = 1;

        boolean moved = false;

        for (int i = 0; i < totalSteps; i++) {
            double newX = x + stepDx;
            double newY = y + stepDy;

            int margin = 2;
            double maxX = WINDOW_WIDTH - w - margin;
            double maxY = WINDOW_HEIGHT - h - margin;
            newX = clamp(newX, margin, maxX);
            newY = clamp(newY, margin, maxY);

            Rectangle testRect = new Rectangle((int) newX, (int) newY, w, h);
            boolean blocked = false;
            for (Wall wall : walls) {
                if (wall.alive && testRect.intersects(wall.getRect())) {
                    blocked = true;
                    break;
                }
            }

            if (!blocked) {
                x = newX;
                y = newY;
                moved = true;
            } else {
                if (dx != 0) {
                    testRect = new Rectangle((int) newX, (int) y, w, h);
                    blocked = false;
                    for (Wall wall : walls) {
                        if (wall.alive && testRect.intersects(wall.getRect())) {
                            blocked = true;
                            break;
                        }
                    }
                    if (!blocked) {
                        x = newX;
                        moved = true;
                    }
                }
                if (dy != 0) {
                    testRect = new Rectangle((int) x, (int) newY, w, h);
                    blocked = false;
                    for (Wall wall : walls) {
                        if (wall.alive && testRect.intersects(wall.getRect())) {
                            blocked = true;
                            break;
                        }
                    }
                    if (!blocked) {
                        y = newY;
                        moved = true;
                    }
                }
                break;
            }
        }

        // 记录痕迹
        if (moved) {
            frameCounter++;
            if (frameCounter % 3 == 0 && alive) {
                TrailPoint tp = new TrailPoint();
                tp.x = this.x + this.w / 2.0;
                tp.y = this.y + this.h / 2.0;
                tp.age = 0;
                trailPoints.add(tp);
            }
        } else {
            frameCounter = 0;
        }
    }

    public Bullet shoot() {
        if (cooldown > 0) return null;
        if (bullets.size() >= MAX_BULLETS) return null;

        cooldown = SHOT_COOLDOWN;
        Utils.Vec2 firePoint = getFirePoint();

        float bulletSpeed = (float)(BULLET_SPEED * getBulletSpeedMultiplier());
        int bulletSize = (int)(BULLET_SIZE * getSizeScale());
        int bulletDamage = getBulletDamage();

        Color bulletColor;
        if (isPlayer) {
            bulletColor = (playerId == 1) ? COLOR_BULLET_P1 : COLOR_BULLET_P2;
        } else {
            bulletColor = COLOR_BULLET_ENEMY;
        }

        Bullet bullet = new Bullet(firePoint.x, firePoint.y, dir, isPlayer, playerId, bulletColor);
        bullet.speed = bulletSpeed;
        bullet.w = bulletSize;
        bullet.h = bulletSize;
        bullet.damage = bulletDamage;
        bullets.add(bullet);
        return bullet;
    }

    public void update(double dt) {
        if (cooldown > 0) cooldown -= dt;

        // 更新道具效果
        for (Iterator<Map.Entry<String, Float>> it = effects.entrySet().iterator(); it.hasNext();) {
            Map.Entry<String, Float> e = it.next();
            e.setValue(e.getValue() - (float)dt);
            if (e.getValue() <= 0) {
                it.remove();
            }
        }

        // 更新痕迹
        for (Iterator<TrailPoint> it = trailPoints.iterator(); it.hasNext();) {
            TrailPoint tp = it.next();
            tp.age += dt;
            if (tp.age >= 1.0) {
                it.remove();
            }
        }

        // 更新子弹
        for (int i = bullets.size() - 1; i >= 0; i--) {
            Bullet b = bullets.get(i);
            b.update(dt);
            if (!b.alive) {
                bullets.remove(i);
            }
        }
    }

    public void draw(Graphics2D g) {
        // ===== 行驶痕迹 =====
        for (TrailPoint tp : trailPoints) {
            double alpha = 60 * (1 - tp.age / 1.0);
            if (alpha > 5) {
                int a = (int) alpha;
                g.setColor(new Color(0, 0, 0, Math.min(a, 255)));
                g.fillRect((int) (tp.x - w / 2.0), (int) (tp.y - h / 2.0), w, h);
            }
        }

        if (!alive) return;

        // ===== Strength 缩放 =====
        float scale = getSizeScale();
        int wDraw = (int)(w * scale);
        int hDraw = (int)(h * scale);
        int xDraw = (int)(x - (wDraw - w) / 2.0);
        int yDraw = (int)(y - (hDraw - h) / 2.0);

        // ===== 主体 =====
        g.setColor(color);
        g.fillRoundRect(xDraw, yDraw, wDraw, hDraw, 6, 6);

        // ===== 边框 =====
        if (isProtected()) {
            g.setColor(Color.WHITE);
            g.setStroke(new BasicStroke(3));
            g.drawRoundRect(xDraw, yDraw, wDraw, hDraw, 6, 6);
            g.setStroke(new BasicStroke(1));
        } else {
            g.setColor(Color.WHITE);
            g.drawRoundRect(xDraw, yDraw, wDraw, hDraw, 6, 6);
        }

        // ===== 炮塔 =====
        Utils.Vec2 center = getCenter();
        int cx = (int) center.x;
        int cy = (int) center.y;
        g.setColor(Color.WHITE);
        g.fillOval(cx - w / 6, cy - w / 6, w / 3, h / 3);

        // ===== 炮管 =====
        int endX = cx + (int) (dir.x * (w / 2.0 + 2));
        int endY = cy + (int) (dir.y * (h / 2.0 + 2));
        g.setStroke(new BasicStroke(4));
        g.drawLine(cx, cy, endX, endY);
        g.setStroke(new BasicStroke(1));

        // ===== 玩家编号 =====
        if (isPlayer) {
            g.setColor(Color.BLACK);
            g.setFont(new Font("Consolas", Font.BOLD, 14));
            String label = String.valueOf(playerId);
            FontMetrics fm = g.getFontMetrics();
            int tw = fm.stringWidth(label);
            g.drawString(label, (int) (x + w / 2 - tw / 2), (int) (y + h / 2 + 5));
        }
    }
}